### ISSUE TITLE
Drag and drop on textarea buggy

### ISSUE THREAD

**Issue by nilshoerrmann** [getkirby/kirby#1281]
**Describe the bug**
When trying to drag images from an image section onto a textarea, the UI breaks. The behaviour across browsers is different: Safari selects the entire document text after dropping, Firefox seems to ignore the dragging except when you drag the image preview directly (not using the drag handle).

**To Reproduce**
Try to drag an section item onto a textarea. Try different browsers.

**Expected behavior**
1. On drag, the UI should always display a ghost item.
2. It should be clear, if you can only drag using the handle or using the image preview as well (the latter does only work in Firefox).
3. On drop, a reference should be inserted into the textarea.

**Screenshots**

This is how it looks like in Safari after dropping:

<img width="1172" alt="grafik" src="https://user-images.githubusercontent.com/25466/50207757-aa01ab00-036f-11e9-96e8-155e880d0763.png">

**Kirby Version**
RC1

**Browsers**
Tested in latest Safari and Firefox versions on macOS Mojave.

**Comment by nilshoerrmann** [comment]
It this interaction currently based on a external library or custom built? I remember there was a discussion about how to handle drag and drop during the alphas but I cannot find it on Slack anymore.

**Comment by bastianallgeier** [comment]
The sorting is done with Sortable.js. Unfortunately that is intervening quite a lot with the regular drag & drop stuff. The sort handle is therefor acting quite differently than the rest of the item.

**Comment by nilshoerrmann** [comment]
Ah, right, thank you. Does it make sense to create a custom Kirby drag & drop script? We often ran into problems with libraries that were solving too many general use cases that we never needed but that cause side effect on the features we actually used. Sounds like the same here.

**Comment by bastianallgeier** [comment]
I have no idea how much effort it really is, but it sounds like a good project in general.

**Comment by nilshoerrmann** [comment]
Maybe this is something we can solve in a community effort? We played around with drag and drop a few times and I was surprised how well the native events work and the panel doesn't support very old browsers, does it?

I really miss the drag and drop option between sections we had in the early alphas. This was a very intuitive interaction that helped solve a common use case.

**Comment by bastianallgeier** [comment]
That would be great. I just don't have the time at the moment, but would love to have a modern alternative. Since we don't support IE at all, it should be pretty cool to work on something that is optimized for modern browsers.

**Comment by nilshoerrmann** [comment]
Alright, let's see what's possible :)

**Comment by bastianallgeier** [comment]
It would maybe make sense to check out Shopify's lib: https://github.com/Shopify/draggable

**Comment by nilshoerrmann** [comment]
Wow, that one is looking great!

**Comment by nilshoerrmann** [comment]
Back to solving the actual issue here: adding image/files to textareas.

If dragging is buggy, would it be possible and maybe easier to add a button to the textarea's toolbar to insert file references, displaying a nice modal that let's the user select from the available files/images?

**Comment by nilshoerrmann** [comment]
Another idea @johannahoerrmann came up with over lunch: add a "copy" option to the image section's settings menu using the [Clipboard API](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API):

![bildschirmfoto am 2018-12-19 um 14 32 27](https://user-images.githubusercontent.com/25466/50223279-0e862f80-039b-11e9-8797-78b677adf389.png)

This would add either a Kirbytext or Markdown reference of the given item to the clipboard so the user could paste it into the textarea. This would at least work in modern browsers.

**Comment by johannahoerrmann** [comment]
A bit of background: This problem was brought to our attention by our client, who needs to rearrange images in their text content. She is solely working in Safari, but as I am working in Firefox and dragged the thumbnails themselves, I didn't notice the problem earlier. Thanks for looking into this! :)

**Comment by nilshoerrmann** [comment]
@bastianallgeier: Any idea of an interim solution we can offer our client? We only need it working in Safari for now.

**Comment by bastianallgeier** [comment]
I will look into it right now.

**Comment by bastianallgeier** [comment]
@nilshoerrmann @johannahoerrmann one question. I can drag images into the textarea as long as I never use the handle. Is that happening for you as well?

**Comment by nilshoerrmann** [comment]
Yes, this works in Firefox for us, too. But that seems to be some kind of native drag and drop, it looks different compared to the sorting dragging.

**Comment by bastianallgeier** [comment]
The native one works in Safari too. Maybe she can use that for now. I think I'd prefer to add the image button to the toolbar, but that's something I can't handle before Christmas.

**Comment by johannahoerrmann** [comment]
> The native one works in Safari too. 

This is a new insight to me! I tried it just now and it works, if you *do not* drag the image (this is what works in Firefox) but the text, and if you have _not tried_ dragging that specific file by its handle before (which results in an entire selection of the site, as seen in @nilshoerrmann's screenshot above). In case of the latter, the native drag and drop doesn't seem to work for this specific file anymore. You need to reload the page first.

Okay, I will pass this on, at least it's something she can get to work!

**Comment by johannahoerrmann** [comment]
However, I see this only as a workaround – in my eyes, this is still a bug and needs fixing. I would expect to be able to drag and drop the entire item, especially using the handle, as this is how I move it within the list.

**Comment by bastianallgeier** [comment]
@johannahoerrmann yes, it's definitely just a pre-christmas "fix"

**Comment by nilshoerrmann** [comment]
I still very much miss drag and drop features that were available in the early alphas of Kirby 3. Now I recently discovered a new drag and drop library on Github that I would like to share here: https://github.com/kutlugsahin/smooth-dnd. Maybe it helps us solving the drag and drop issues present in the Panel? The library comes with a specific Vue wrapper: https://github.com/kutlugsahin/vue-smooth-dnd/.

I've not use the library myself but the demos look indeed very smooth:
https://kutlugsahin.github.io/vue-smooth-dnd/

**Comment by johannahoerrmann** [comment]
Definitely missing this, too! What came up as a "solution" in this discussion was only considered a pre-christmas "fix" by @bastianallgeier.

**Comment by distantnative** [comment]
@johannahoerrmann I mean... it's still kinda pre-christmas, isn't it? ;) I can see Christmas cookies in my local supermarkt :D

**Comment by distantnative** [comment]
Back to serious: We will have a look at that library. It might fix some of the issues. Others are more conceptually that allowing drag and drop just opens up a whole new world of problems (not specifically in regard to this issue, more to the broad dnd features we had in early alphas).

**Comment by johannahoerrmann** [comment]
@distantnative Shockingly, you are absolutely right! It _is_ pre-christmas again! 
🎅🏽

**Comment by nilshoerrmann** [comment]
@distantnative Maybe fixing drag and drop could be a task for a specific minor release? The drag and drop behaviour of the alphas really made the interface stand out and made it feel more natural and native (in the sense of belonging to the operating system).

**Comment by distantnative** [comment]
@nilshoerrmann Considering all the work and implications, I think it's more a mejor release on its own to be honest

**Comment by nilshoerrmann** [comment]
Major like Kirby 4?! I was thinking in terms of `major.minor.patch`, something like Kirby 3.5.

**Comment by bastianallgeier** [comment]
I think @distantnative means a minor release, but we think of them more as feature releases. So more like `major.feature.minor` or `major.feature.path`

**Comment by nilshoerrmann** [comment]
Alright, we're in the same boat then. :+1:

**Comment by bastianallgeier** [comment]
You know what makes those issues so hard? In order to fix them you really have to know the details of the drag & drop implementation and it's even better to actually have written your own lib in order to completely understand it. I started that a couple months ago, but it's a nightmare. What we have right now is a drag & drop library, that works in some places and doesn't in others and it interfers with native drag & drop and it's all a mess. But to get this all sorted we need to invest weeks of work and we can't just put in another library and hope that it will fix it. We also would need to do this as a team. It doesn't make sense if I write a library that I understand, but @distantnative or @lukasbestle are not on board and whenever something breaks they don't understand it. This is really hard, because the drag & drop API is not very easy to work with in the first place. We have to become drag & drop experts before we can solve this. 

It's a bit like the dilemma with the markdown parser. It's all great until you run into issues, but once you run into issues with the parser you are screwed. That's exactly the reason why I'm so careful with dependencies. But in some places we have to use them in order to get stuff done that we are not experienced enough to write ourselves.

**Comment by bastianallgeier** [comment]
So to make it short: the amount of work and self-education that we need to throw at these problems is completely out of proportions compared to the urgency of the problems.

**Comment by distantnative** [comment]
@nilshoerrmann oh sorry, no not like Kirby 4... was thinking of `3.major.minor` - sorry for the shock :D

**Comment by nilshoerrmann** [comment]
> So to make it short: the amount of work and self-education that we need to throw at these problems is completely out of proportions compared to the urgency of the problems.

Oh, I politely disagree :) 

Yes, of course this is a complex topic and yes, there is no easy solution. But regarding the urgency, this is only true when you are focusing on the code part of Kirby. When looking at the Panel as a end user, not having a working drag and drop solution complicates the interactions a lot because I have to find options hidden in dropdowns. Drag and drop is about intuitiveness. And drag and drop is something very natural these days, users try to drag and drop. Yes, other systems fail here as well. But that's why users often consider using a CMS complicated.

So making the interface behave the way the end users expect it is very important. It will make the CMS stand out. And thus it's as urgent as other issues in my eyes. The difference to other issues is that it's about fixing behaviour and not about fixing logic.

**Comment by distantnative** [comment]
You might have misread. Saying that the effort-urgency ratio is out of proportion, does not mean that is is neither important nor urgent. Bastian just explained what a huge effort this needs - this is what ruins the proportions, not that it is not important. It is less a direct matter of priorities than being able to get work done with our current knowledge, experience etc.

So in a basket of things that are all important and urgent, we need to start with the things we can get done and try to learn more to become able to get those other things done eventually as well.

**Comment by bastianallgeier** [comment]
@distantnative you explained it far better than me :) Exactly this! I agree that it's as important as the rest, but we are limited to the team size and the knowledge that we have right now. If we could throw money at it, I'd probably try to find someone who has a lot of experience with drag & drop implementations before and let them work on it, but with 1-3 people working on it, we have to be extremely practical sometimes –  if we like it or not. 

Another good example for this is the WYSIWYG editor. It took me almost two years with lots of experiments on the side while working on v3 to get to the point where the editor is now.

**Comment by afbora** [comment]
@nilshoerrmann About 2 years have passed. I wonder if this issue still persists. Because it's not macOS, but I haven't seen a issue in Windows/Firefox 🤔

**Comment by nilshoerrmann** [comment]
As far as I'm aware nothing has changed. But I'll have to check, especially the Safari behaviour.

**Comment by nilshoerrmann** [comment]
@afbora: I just checked an the issue persist across browsers on macOS. Things seems to have become better in Safari but in general the issue still exists.

**Comment by distantnative** [comment]
I don't see us finding a fix for this if the underlying library won't get fixed itself.
I could see a "Copy reference" dropdown option that gets then pasted as Kirbytag or Markdown as the solution we should consider and close this then.

**Comment by bastianallgeier** [comment]
We are really sorry that we didn't manage to solve this issue for such a long time. Unfortunately, we currently don't see a possible solution and will close this issue for now. Please create a new issue if this is still an urgent problem that affects your work/projects.
